packages/core/src/browser/opener-service.ts

```diff
@@ -84,12 +88,27 @@ export async function open(openerService: OpenerService, uri: URI, options?: Ope
 
 @injectable()
 export class DefaultOpenerService implements OpenerService {
+    // Collection of open-handlers for custom-editor contributions.
+    protected readonly customEditorOpenHandlers: OpenHandler[] = [];
+
+    protected readonly onOpenersStateChangedEmitter = new Emitter<void>();
+    readonly onOpenersStateChanged = this.onOpenersStateChangedEmitter.event;
 
     constructor(
         @inject(ContributionProvider) @named(OpenHandler)
         protected readonly handlersProvider: ContributionProvider<OpenHandler>
     ) { }
 
+    public addHandler(openHandler: OpenHandler): Disposable {
```

Comment:
Is there a reason we do not want to add `addHandler` to the `OpenerService` interface.

The method is only called by the plugin system, shouldn't the plugin side implement a `CustomEditorOpenerService` who's responsibility is solely on registering and providing handlers for the OpenerService on the plugin side rather than trying to fit it in the `Default`?

Reply:
> Is there a reason we do not want to add addHandler to the OpenerService interface.

Added it.

>The method is only called by the plugin system, shouldn't the plugin side implement a CustomEditorOpenerService who's responsibility is solely on registering and providing handlers for the OpenerService on the plugin side rather than trying to fit it in the Default?

I'm not sure I follow: 
a) Method is called when `CustomEditor` contributions are loaded and used to register CustomEditor handlers in `DefaultOpenerService`. 
b) Opening file triggers `OpenerService` `open` method which retrieves all handlers to find relevant one for that file type.

You suggest (a) would register handlers in new `CustomerEditorOpenerService`? and how would we retrieve relevant handler on opening file?

Reply:
@danarad05 my main concern is that it breaks the current contract of how the framework contributes `OpenHandlers`. With these changes we now add the ability to `addHandler` which is only used by the plugin-system but is not something we necessarily want `@theia` extensions to call. In addition, the current code seems to handle only when custom-editors are registered, but do we need to cleanup when plugins are removed (we may end up in state where a removed plugin still has the highest `canHandle` opener).

I'm not blocking the pull-request, only expressing concerns with the current approach on the framework side.

Reply:
@vince-fugnitto i think that when plugin is removed its disposable collection is applied including lines https://github.com/eclipse-theia/theia/pull/8910/files/5d62c8be0987b29240f0d29268ea38b9a9c08502#diff-189a4e3822a82cd7361ed8bc78201cfd03cb707e5e50d30f3567a172ad2975e4R107-R108 below.

I think this `addHandler` API can also be used from theia extension dynamically. Not only from plugin.

Reply:
> my main concern is that it breaks the current contract of how the framework contributes OpenHandlers. With these changes we now add the ability to addHandler which is only used by the plugin-system but is not something we necessarily want @theia extensions to call. 

I'm not objecting, I'm just not sure how to implement it otherwise - if you know please let me know.

>In addition, the current code seems to handle only when custom-editors are registered, but do we need to cleanup when plugins are removed (we may end up in state where a removed plugin still has the highest canHandle opener).

As @amiramw stated, dispose of CustomEditorWidget entails removal of opener - tested to verify.